core/src/main/java/com/linecorp/armeria/server/decoding/DecodingService.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.decoding;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+import java.util.Map;
+import java.util.function.Function;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.client.encoding.StreamDecoderFactory;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.server.DecoratingService;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.SimpleDecoratingHttpService;
+
+/**
+ * A {@link DecoratingService} that requests and decodes HTTP encoding (e.g., gzip) that has been applied to the
+ * content of an {@link HttpRequest}.
+ */
+public class DecodingService extends SimpleDecoratingHttpService {
+
+    /**
+     * Creates a new {@link DecodingService} decorator with the default encodings of 'gzip' and 'deflate'.
+     */
+    public static Function<? super HttpService, DecodingService> newDecorator() {
```

Comment:
Users might want to customize `StreamDecoderFactory`s.
The following factories would be useful for that case.
https://github.com/line/armeria/blob/eb33d2b77fb9f600b3eee14d4c601fdef32add6f/core/src/main/java/com/linecorp/armeria/client/encoding/DecodingClient.java#L51-L62

By the way, We need to discuss `StreamDecoderFactory` issue before going further.

Reply:
In my opinion, `StreamDecoderFactory`, `StreamDecoderFactories` and `ZlibStreamDecoder` package need to move from `com.linecorp.armeria.client.encoding` to `com.linecorp.armeria.common` because roles are not tied to the client. but this action bring many class changes.

Reply:
To avoid breaking changes, how about?
1. Deprecate `client.encoding`
2. Copy the code to `common.encoding`

Reply:
That's a good idea. Can I do that?

Reply:
Sure! Please do that. 🙏

Reply:
I just finished it and pushed.